File: Unidad2/MecanismosOrdenacion/MergeSort.py
def merge_sort(lista):

    if len(lista) <= 1:
        return lista
    medio = len(lista) // 2

    izquierda = merge_sort(lista[:medio])
    derecha = merge_sort(lista[medio:])

    resultado = []
    i = j = 0

    while i < len(izquierda) and j < len(derecha):
        if izquierda[i] < derecha[j]:
            resultado.append(izquierda[i])
            i += 1
        else:
            resultado.append(derecha[j])
            j += 1

    resultado.extend(izquierda[i:])
    resultado.extend(derecha[j:])

    return resultado
numeros = [8, 3, 1, 7, 4]
print(merge_sort(numeros))

File: Unidad2/MecanismosOrdenacion/MergeSort.java
import java.util.Arrays;
public class MergeSort {

    static int[] mergeSort(int[] lista) {

        if (lista.length <= 1)
            return lista;
        int medio = lista.length / 2;

        int[] izquierda = mergeSort(Arrays.copyOfRange(lista, 0, medio));
        int[] derecha = mergeSort(Arrays.copyOfRange(lista, medio, lista.length));

        int[] resultado = new int[lista.length];
        int i = 0, j = 0, k = 0;

        while (i < izquierda.length && j < derecha.length) {
            if (izquierda[i] < derecha[j])
                resultado[k++] = izquierda[i++];
            else
                resultado[k++] = derecha[j++];
        }

        while (i < izquierda.length)
            resultado[k++] = izquierda[i++];

        while (j < derecha.length)
            resultado[k++] = derecha[j++];

        return resultado;
    }
    public static void main(String[] args) {
        int[] numeros = {8, 3, 1, 7, 4};
        numeros = mergeSort(numeros);

        System.out.println(Arrays.toString(numeros));
    }
}